### PR TITLE
Script Resources and Overview PDFs

### PR DESCRIPTION
**Pull request by Hamms**
Add [Script Overview](https://github.com/code-dot-org/code-dot-org/files/6268163/CSD.Unit.1.-.Problem.Solving.and.Computing.21-.22.pdf) and [Script Resources](https://cdo-lesson-plans-dev.s3.amazonaws.com/csp1-2021/20210324214035/CSP+Unit+1+-+Digital+Information+('21-'22)+-+Resources.pdf) pdfs to the existing PDF generation process.

## Follow-up work

- There's a potential optimization we could add to retrieve lesson PDFs from S3 if they exist. It's pretty unnecessary right now; because we're generating the overview PDF as part of a process which also generates the lesson PDFs, we expect that the PDFs will always be available in the local filesystem. But there will probably be scenarios in the future where we'd like to generate one without the other, so pulling from S3 would be nice.
- Some Resources aren't being retrieved correctly from Google right now, and will need their URLs tweaked in order to do so
  - `csp_unit_1_-_digital_information` has the url https://docs.google.com/presentation/d/137WkLc4Lgld2cYrDrru-qAunb58S59GByTl4TpXWnrs/copy rather than https://docs.google.com/presentation/d/137WkLc4Lgld2cYrDrru-qAunb58S59GByTl4TpXWnrs/edit, for example

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by dmcavoy** [comment]
The example for script overview looks like it is getting the student view of the unit. Is that intentional? The one I worked on was for the teacher view with the calendar and lesson group description/questions included.

**Comment by Hamms** [comment]
We're only capable of printing to PDF the signed-out view for any given page, not a user-specific view. If that's a requirement, we're going to have to figure out a way to present the "teacher" view outside of the context of a specific user

**Comment by dmcavoy** [comment]
I do think long term we want to print what the teacher sees such as the teacher overview (the signed out overview would be the student one I assume). @tess323 to weigh in here. There are some quick wins here to get the signed out page to show what we want for now.

If you take away the viewtype requirement here: https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/code-studio/components/progress/ScriptOverview.jsx#L155 we should be able to show the calendar I would assume.

We should hide the Lesson Resources button. 

Can you check that a unit with a lesson group description and big questions will show on the signed out version of the print? I think it will from my quick look but it would be good to check.

**Comment by Hamms** [comment]
SGTM! I'll open a separate PR with those changes.

> Can you check that a unit with a lesson group description and big questions will show on the signed out version of the print?

Yep, they show up [CSD Unit 1 - Problem Solving and Computing ('21-'22).pdf](https://github.com/code-dot-org/code-dot-org/files/6245477/CSD.Unit.1.-.Problem.Solving.and.Computing.21-.22.pdf)

**Comment by Hamms** [comment]
Do we also want the expandable lesson sections to be expanded?

**Comment by tess323** [comment]
Agreed with Dani that we will want to match the teacher view for this as much as is possible. 

- Unit Overview calendar would be a great quick win
- If it were possible to get the teacher overview that would be another great win

Yes as well to expanding Lesson Sections

File: dashboard/lib/services/curriculum_pdfs.rb
Comment (dmcavoy) [inline dashboard/lib/services/curriculum_pdfs.rb:10]: Do we also want to include the course overview in the all lessons PDF when you get it from the course page?

Comment (dmcavoy) [reply dashboard/lib/services/curriculum_pdfs.rb:10]: I guess maybe said a different way it looks like this adds the script overview pdf. Do we need a unit group overview (course overview) pdf which is a combo of the script overview pdfs and the course overview page pdf?

**Comment by Hamms** [comment]
This discussion would I think benefit from stepping back a bit and clarifying a few broader issues.

Right now, the script overview defaults to the student view, and the student view behaves differently from the teacher view in a number of ways. It seems like for these PDFs, we want to at the very least have what we generate be closer to the teacher view, if not identical to it. So, some potential options, presented in the order of my preference:

1. We could change the page to show the teacher view to signed-out users by default, rather than the student view
2. We could update the student view to look more like the teacher view (calendar, expanded sections)
3. We could add a url parameter that would force the page into teacher view rather than student view

Thoughts?

**Comment by davidsbailey** [comment]
> This discussion would I think benefit from stepping back a bit and clarifying a few broader issues.
> 
> Right now, the script overview defaults to the student view, and the student view behaves differently from the teacher view in a number of ways. It seems like for these PDFs, we want to at the very least have what we generate be closer to the teacher view, if not identical to it. So, some potential options, presented in the order of my preference:
> 
> 1. We could change the page to show the teacher view to signed-out users by default, rather than the student view
> 2. We could update the student view to look more like the teacher view (calendar, expanded sections)
> 3. We could add a url parameter that would force the page into teacher view rather than student view
> 
> Thoughts?

This seems like a good set of options. I don't mean to be contrary, but option 3 seems the best to me from a "do no harm" perspective. I think the other options would require product buy-in, in which case those could be fine too. The reason I feel this way is that I wouldn't want the logistics of PDF generation to dictate what a student or signed-out user sees by default on our site, however there could very well be room for a convergence of interests here.

**Comment by davidsbailey** [comment]
@tess323 perhaps you can weigh in on options 1-3 from a product perspective?

**Comment by tess323** [comment]
Would option 3 - adding the URL perimeter allow the teacher and student view to stay the same?  If so that’s ideal especially if it is not significantly more work.

we will just want to ensure the verified teacher content (resources) is not showing. Otherwise I will need to follow up with PL.

**Comment by dmcavoy** [comment]
Screen Shot 2021-04-03 at 7 50 13 AM

There is a url param that shows up when you switch between student and teacher view using the teacher panel. Maybe we can use that?

**Comment by davidsbailey** [comment]
CC @jamescodeorg since this affects script overview which I know LP has been working on lately.

**Comment by jamescodeorg** [comment]
Thanks for the tag!

Are there any concerns with the teacher materials being (easily) visible to students?

I don't have as much context and you all do, but it feels like driving changes into this view for signed-out users because of pdf generation is the wrong general direction.  If we generally want our pdfs to reflect what a teacher sees, maybe we should have the pdf generation code look like a teacher account to the rest of the system?

**Comment by tess323** [comment]
Existing behavior on Curriculum Builder is lesson plans are available to signed-out users.  There is content hidden behind a login for verified teachers - namely restricted resources.  The student overview is a new feature, so this change would not be straying from current functionality.  

That said, super open to Jame's suggestion above if it were to play well with our timeline -- this feature will need to be widely used on 4/23.

**Comment by Hamms** [comment]
PTAL; script overview should now be using the teacher view: [CSD Unit 1 - Problem Solving and Computing ('21-'22).pdf](https://github.com/code-dot-org/code-dot-org/files/6268167/CSD.Unit.1.-.Problem.Solving.and.Computing.21-.22.pdf)

**Comment by davidsbailey** [comment]
> * Some Resources aren't being retrieved correctly from Google right now, and will need their URLs tweaked in order to do so

This is needed for April 23, right? Will this be done as part of [PLAT-759], or needs to be tracked separately?

[PLAT-759]: https://codedotorg.atlassian.net/browse/PLAT-759

File: dashboard/lib/services/curriculum_pdfs.rb
Comment (dmcavoy) [reply dashboard/lib/services/curriculum_pdfs.rb:10]: @tess323 Any thoughts on this?

**Comment by dmcavoy** [comment]
> CSD Unit 1 - Problem Solving and Computing ('21-'22).pdf

It looks like the locked lesson bubbles don't show. Is this an issue? @tess323

File: dashboard/lib/services/curriculum_pdfs.rb
Comment (tess323) [reply dashboard/lib/services/curriculum_pdfs.rb:10]: @dmcavoy - Looking at CB we do not currently have this functionality so it doesn't feel like P0 to me. That said I think it makes sense in theory to have something like this if we could get it for cheap.  My priorities are:

- getting PDFs that roughly match curriculum builder live so we can run them by PL 
- adding any additional PDFs we think we can sneak in under the wire

**Comment by tess323** [comment]
@dmcavoy - its definitely not ideal, but not sure if its a ship blocker as showing these levels on this page is a new feature.  I assume this is because the default is locked for signed out users.  Is the default unlocked for signed in users? We wouldn't want students to be able to access locked levels on signed out pages. One idea could we add default text for when there are locked levels saying "There are locked levels in this lesson".  Another is we could default show levels on lockable lessons, but just make sure you cannot access them (which I believe is the case for all levels logged out in CSP and CSD).  Any other options I may be missing?

**Comment by Hamms** [comment]
I'm not familiar with what these locked/unlocked lesson bubbles are; can you enlighten me?

**Comment by Hamms** [comment]
I would note that feedback on the specifics of how the script overview page is rendered into a PDF should be associated more with the "make script overview page able to be rendered into a PDF" work item, not this, which is the "assemble PDFs from resources and existing page PDFs" work item. I'm happy to own any followup work that needs to be done for the former, but none of that should block this PR.

**Comment by Hamms** [comment]
> This is needed for April 23, right? Will this be done as part of PLAT-759, or needs to be tracked separately?

This should be done as follow-up work. I'm tracking it at https://codedotorg.atlassian.net/browse/PLAT-952

**Review by davidsbailey — APPROVED** [review]
the code here looks good to me. nice job getting this all working! LGTM once other people's concerns are addressed.

**Comment by tess323** [comment]
> I'm not familiar with what these locked/unlocked lesson bubbles are; can you enlighten me?

I think @dmcavoy has the deepest context here.  A lesson can be marked as lockable, which primarily is a tool for giving assessments.  A teacher can lock and unlock a lesson based on where students are in the assessment-taking process.  You can play around with it on CSD Unit 1 Lesson 1

That said - I agree this shouldn't block this PR which is the basic "assemble PDFs from resources and existing page PDFs" functionality this PR is aiming to accomplish.

**Pull request merged**
